R/compiler-flags.R

```diff
@@ -16,22 +16,36 @@
 #' compiler_flags()
 #' compiler_flags(TRUE)
 compiler_flags <- function(debug = FALSE) {
-  if (Sys.info()[["sysname"]] == "SunOS") {
+  res <-
+    if (Sys.info()[["sysname"]] == "SunOS") {
     c(
       CFLAGS   = "-g",
-      CXXFLAGS = "-g"
+      CXXFLAGS = "-g",
+      CXX11FLAGS = "-g"
     )
   } else if (debug) {
     c(
-      CFLAGS   = "-UNDEBUG -Wall -pedantic -g -O0 -fdiagnostics-color=always",
-      CXXFLAGS = "-UNDEBUG -Wall -pedantic -g -O0 -fdiagnostics-color=always",
+      CFLAGS   = "-UNDEBUG -Wall -pedantic -g -O0",
+      CXXFLAGS = "-UNDEBUG -Wall -pedantic -g -O0",
+      CXX11FLAGS = "-UNDEBUG -Wall -pedantic -g -O0",
       FFLAGS   = "-g -O0",
       FCFLAGS  = "-g -O0"
     )
   } else {
     c(
-      CFLAGS   = "-Wall -pedantic -fdiagnostics-color=always",
-      CXXFLAGS = "-Wall -pedantic -fdiagnostics-color=always"
+      CFLAGS   = "-Wall -pedantic",
+      CXXFLAGS = "-Wall -pedantic",
+      CXX11FLAGS = "-Wall -pedantic"
     )
   }
+
+  if (has_compiler_colored_diagnostics()) {
+    res[c("CFLAGS", "CXXFLAGS", "CXX11FLAGS")] <-
+      paste(res[c("CFLAGS", "CXXFLAGS", "CXX11FLAGS")], "-fdiagnostics-color=always")
+  }
+  res
+}
+
+has_compiler_colored_diagnostics <- function() {
```

Comment:
We could cache this, I think.

Reply:
It takes ~700 _microseconds_ to run on average, don't think we need to cache it. It is essentially the same speed as `crayon::has_color()` takes.

```
> microbenchmark::microbenchmark(has_compiler_colored_diagnostics())
Unit: microseconds
                               expr     min      lq    mean  median      uq
 has_compiler_colored_diagnostics() 688.547 729.437 821.151 779.252 838.361
      max neval
 2706.717   100
> microbenchmark::microbenchmark(crayon::has_color())
Unit: microseconds
                expr     min      lq    mean  median      uq      max neval
 crayon::has_color() 604.295 620.662 669.354 653.316 675.115 1081.829   100
```

Reply:
I am pretty sure that it takes more on windows, because starting processes is slow on windows.

Btw. `crayon::has_color()`'s speed is different in different environments, e.g. in the terminal it is slowish:
```
❯ microbenchmark::microbenchmark(crayon::has_color())
Unit: microseconds
                expr     min       lq     mean  median      uq     max neval
 crayon::has_color() 212.679 215.1865 226.1735 218.678 224.765 342.662   100
```

But it is much faster in RStudio, because it does not have to query much:
```
> microbenchmark::microbenchmark(crayon::has_color())
Unit: microseconds
                expr    min     lq     mean  median     uq   max neval
 crayon::has_color() 14.268 14.699 18.50084 15.0105 19.095 83.22   100
```

Reply:
But yeah, if is is less than (say) 200 ms, then we don't need to cache it.

Reply:
Although it is not hard to cache, just using a closure.

Reply:
Fine, I cached it, It will be potentially wrong if you switch compilers without restarting R, but maybe worthwhile.

Reply:
Turns out there was built-in machinery for caching already in pkgbuild, so I have switched to using that. I also did not realize that `has_compiler()` was already cached, so needed to disable it for this case.

Reply:
Switching compilers without restarting R is usually a bad idea, I think. The best is to use the same compiler R was compiled with....